From e16a712fac53f82ba0f0a6faf9b6f6f271ac31e3 Mon Sep 17 00:00:00 2001
From: David Gibson <dgibson@redhat.com>
Date: Mon, 15 Sep 2014 07:12:39 -0400
Subject: [virt] kvm/ppc: book3s/hv - Don't drop low-order page address bits

Message-id: <1410765214-16377-24-git-send-email-dgibson@redhat.com>
Patchwork-id: 94725
O-Subject: [PATCH 23/78] KVM: PPC: Book3S HV: Don't drop low-order page address bits
Bugzilla: 1123145 1123133 1123367
RH-Acked-by: Radim Krcmar <rkrcmar@redhat.com>
RH-Acked-by: Don Zickus <dzickus@redhat.com>
RH-Acked-by: Steve Best <sbest@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1123145

Signed-off-by: David Gibson <dgibson@redhat.com>

commit df9059bb64023da9f27e56a94a3e2b8f4b6336a9
Author: Paul Mackerras <paulus@samba.org>
Date:   Mon Dec 16 13:31:46 2013 +1100

    KVM: PPC: Book3S HV: Don't drop low-order page address bits

    Commit caaa4c804fae ("KVM: PPC: Book3S HV: Fix physical address
    calculations") unfortunately resulted in some low-order address bits
    getting dropped in the case where the guest is creating a 4k HPTE
    and the host page size is 64k.  By getting the low-order bits from
    hva rather than gpa we miss out on bits 12 - 15 in this case, since
    hva is at page granularity.  This puts the missing bits back in.

    Reported-by: Alexey Kardashevskiy <aik@ozlabs.ru>
    Signed-off-by: Paul Mackerras <paulus@samba.org>
    Signed-off-by: Alexander Graf <agraf@suse.de>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/powerpc/kvm/book3s_hv_rm_mmu.c b/arch/powerpc/kvm/book3s_hv_rm_mmu.c
index 693ba77..ff56885 100644
--- a/arch/powerpc/kvm/book3s_hv_rm_mmu.c
+++ b/arch/powerpc/kvm/book3s_hv_rm_mmu.c
@@ -241,6 +241,7 @@ long kvmppc_do_h_enter(struct kvm *kvm, unsigned long flags,
    is_io = hpte_cache_bits(pte_val(pte));
    pa = pte_pfn(pte) << PAGE_SHIFT;
    pa |= hva & (pte_size - 1);
+   pa |= gpa & ~PAGE_MASK;
   }
  }
 
-- 
1.7.1